components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiments.kt

```diff
@@ -146,35 +140,100 @@ open class ExperimentsInternalAPI internal constructor() {
+            findAndStartActiveExperiment()
+        }
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun findAndStartActiveExperiment() {
+        assert(activeExperiment == null) { "Should not have an active experiment" }
+
+        evaluator.findActiveExperiment(context, experimentsResult.experiments)?.let {
+            logger.info("""Activating experiment - id="${it.experiment.id}", branch="${it.branch}"""")
+            activeExperiment = it
+            it.save(context)
+            Glean.setExperimentActive(it.experiment.id, it.branch)
+        }
+    }
+
+    private fun stopActiveExperiment() {
+        assert(activeExperiment != null) { "Should have an active experiment" }
+
+        ActiveExperiment.clear(context)
+        activeExperiment = null
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun loadActiveExperiment(
+        context: Context,
+        experiments: List<Experiment>
+    ): ActiveExperiment? {
+        assert(activeExperiment == null) { "Should not have an active experiment" }
+
+        val activeExperiment = ActiveExperiment.load(context, experiments)
+        logger.info(activeExperiment?.let
+        { """Loaded active experiment from cache - id="${it.experiment.id}", branch="${it.branch}"""" }
+            ?: "No active experiment in cache"
+        )
+
+        return activeExperiment
     }
 
     /**
      * Checks if the user is part of
      * the specified experiment
      *
-     * @param context context
      * @param experimentId the id of the experiment
+     * @param branchName the name of the branch for the experiment
      *
      * @return true if the user is part of the specified experiment, false otherwise
      */
     @VisibleForTesting(otherwise = VisibleForTesting.NONE)
-    internal fun isInExperiment(context: Context, experimentId: String): Boolean {
-        return evaluator.evaluate(context, ExperimentDescriptor(experimentId), experimentsResult.experiments) != null
+    internal fun isInExperiment(experimentId: String, branchName: String): Boolean {
+        return activeExperiment?.let {
+            (it.experiment.id == experimentId) &&
+            (it.branch == branchName)
+        } ?: false
     }
 
     /**
      * Performs an action if the user is part of the specified experiment
      *
-     * @param context context
      * @param experimentId the id of the experiment
      * @param block block of code to be executed if the user is part of the experiment
      */
-    fun withExperiment(context: Context, experimentId: String, block: (branch: String) -> Unit) {
-        val activeExperiment = evaluator.evaluate(
-            context,
-            ExperimentDescriptor(experimentId), experimentsResult.experiments
-        )
-        activeExperiment?.let { block(it.branch) }
+    fun withExperiment(experimentId: String, block: (branch: String) -> Unit) {
```

Comment:
If we make this inline, it then exposes some of the internal variables through the public API.  I'll file a follow-up bug to investigate if we can reduce the memory overhead here and other places, but it looks like it would require more refactoring to accomplish.

Reply:
👍